lib/std/heap/general_purpose_allocator.zig

```diff
@@ -1397,6 +1403,11 @@ test "double frees" {
     try std.testing.expect(GPA.searchBucket(&gpa.buckets[index], @intFromPtr(small.ptr), gpa.cur_buckets[index]) == null);
     try std.testing.expect(GPA.searchBucket(&gpa.empty_buckets, @intFromPtr(small.ptr), null) != null);
 
+    // the metadata tracking system relies on alloc_cursor of empty buckets
+    // being set to the slot count so that we can get back the size class.
+    const empty_bucket = GPA.searchBucket(&gpa.empty_buckets, @intFromPtr(small.ptr), null).?;
+    try std.testing.expect(@divExact(page_size, empty_bucket.alloc_cursor) == size_class);
+
```

Comment:
Not sure if this test should be included in this PR, but I found it not obvious that this was the way to extract the `size_class` from an empty bucket. It's hidden away in a branch of the `free` function and if this invariant changes the metadata tracking system will break.

Reply:
How about adding a function to `BucketHeader`, something like:

```zig
/// Only valid for buckets within `empty_buckets`, and relies on the `alloc_cursor`
/// of empty buckets being set to `slot_count` when they are added to `empty_buckets`
fn sizeClassOfEmptyBucket(bucket: *BucketHeader) usize {
    // alloc_cursor was set to slot count when bucket added to empty_buckets
    // so we can get the size_class
    return @divExact(page_size, bucket.alloc_cursor);
}
```

and then using that everywhere this `@divExact(page_size, bucket.alloc_cursor)` calculation is done?